File: WindowsFormsApplication1/WindowsFormsApplication1/Boat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using System.Windows.Forms;

namespace WindowsFormsApplication1
{
    public class Boat : Vehicle
    {
        public override int MaxSpeed
        {
            get
            {
                return base.MaxSpeed;
            }

            protected set
            {
                if (value > 0 && value < 100)
                {
                    base.MaxSpeed = value;
                }
                else
                {
                    base.MaxSpeed = 50;
                }
            }
        }
        public override int MaxCountPassengers
        {
            get
            {
                return base.MaxCountPassengers;
            }

            protected set
            {
                if (value > 0 && value < 150)
                {
                    base.MaxCountPassengers = value;
                }
                else
                {
                    base.MaxCountPassengers = 140;
                }
            }
        }
        public override double Weight
        {
            get
            {
                return base.Weight;
            }

            protected set
            {
                if (value > 500 && value < 1500)
                {
                    base.Weight = value;
                }
                else
                {
                    base.Weight = 1000;
                }
            }
        }
        public Boat(int maxSpeed, int maxCountPassenger, double weight, Color color)
        {
            this.MaxSpeed = maxSpeed;
            this.MaxCountPassengers = maxCountPassenger;
            this.ColorBody = color;
            this.Weight = weight;
            this.countPassengers = 0;
            Random rand = new Random();
            startPosX = rand.Next(10, 200);
            startPosY = rand.Next(10, 200);
        }
        public override void moveBoat(Graphics g)
        {
            startPosX +=
                (MaxSpeed * 50 / (int)Weight) /
                (countPassengers == 0 ? 1 : countPassengers);
            drawBoat(g);
        }
        public override void drawBoat(Graphics g)
        {
            drawLightBoat(g);
        }
        public void FillPolygonPoint(PaintEventArgs e)
        {
            SolidBrush br = new SolidBrush(ColorBody);
            Point point1 = new Point(startPosX + 40, startPosY + 20);
            Point point2 = new Point(startPosX + 40, startPosY);
            Point point3 = new Point(startPosX + 60, startPosY);
            Point[] curvePoints = { point1, point2, point3 };
            e.Graphics.FillPolygon(br, curvePoints);
        }
        protected virtual void drawLightBoat(Graphics g)
        {
            Pen pen = new Pen(ColorBody, 5);
            Point[] a = { new Point(startPosX, startPosY + 20), new Point(startPosX + 60, startPosY + 20), new Point(startPosX + 40, startPosY + 35), new Point(startPosX, startPosY + 35) };
            g.DrawPolygon(pen, a);
        }
    }
}

File: TP_laba2/src/Boat.java
import java.awt.*;

public class Boat extends Vehicle {

	public Boat(int maxSpeed, int maxCountPassengers, int weight, Color color) {
		setMaxSpeed(maxSpeed);
		setMaxCountPassengers(maxCountPassengers);
		setColorBody(color);
		setWeight(weight);
		countPassengers = 0;

		startPosX = 75;
		startPosY = 90;
	}

	@Override
	protected void setMaxSpeed(int value) {
		if (value > 0 && value < 300) {
			super.setMaxSpeed(value);
		} else {
			super.setMaxSpeed(150);
		}
	}
	@Override
	protected void setMaxCountPassengers(int value) {
		if (value > 0 && value < 16) {
			super.setMaxCountPassengers(value);
		} else {
			super.setMaxCountPassengers(15);
		}
	}

	

	@Override
	public void setWeight(int value) {
		if (value > 500 && value < 1500) {
			super.setWeight(value);
		} else {
			super.setWeight(1000);
		}
	}

	

	
	@Override
	public void moveVehicle() {
		startPosX += (int) (getMaxSpeed() * 50 / getWeight() / (countPassengers == 0 ? 1 : countPassengers));
	}

	@Override
	public void drawVehicle(Graphics g) {
		drawBoat(g);
	}

	protected void drawBoat(Graphics g) {
g.setColor(getColorBody());
		
		int x = startPosX;
		int y = startPosY;
		int[] xPoints = new int[] { x, x + 60, x + 40, x, x };
		int[] yPoints = new int[] { y, y, y + 20, y + 20, y };
		g.fillPolygon(xPoints, yPoints, xPoints.length);

	}

	

}
